### PR TITLE
Improve `elastic-agent install` performance

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR makes `elastic-agent install` run slightly faster by concurrently copying files when the Agent is being installed on a host with SSDs.  The same is done with Elastic Agent upgrades.

### Summary of installation times

For all tests, the first installation result was discarded and the worse of the next two installation results (the result with the higher installation time) was used.

| OS | Disk type | Before this PR (s) | After this PR (s) | Time savings (s) | Time savings (%) |
| --- | --------- | ----------------------------- | --------------------------- | ----------------- | ----------------- |
| MacOS | SSD | 15.3s | 13.9s | 1.4s | 9.2% |
| MacOS | HDD | | | | |
| Linux | SSD | 12.2s | 9.2s | 3s | 24.6% |
| Linux | HDD | 45.7s | 46s | -0.3s | -0.7% |
| Windows | SSD | 9.6s | 6.3s | 3.3s | 34.4% |
| Windows | HDD | | | | |

### MacOS - SSD
Tested on 2023 MacOS (Ventura 13.5) running SSDs. 

#### Before this PR
```
$ time sudo ./elastic-agent install -n
Installing in non-interactive mode.Uninstalling current Elastic Agent... DONE
Copying files............................................................. DONE
Installing service... DONE
Starting service... DONE
Elastic Agent has been successfully installed.
sudo ./elastic-agent install -n  0.51s user 2.90s system 22% cpu 15.265 total
$ find . | wc -l
    3464
$ sudo find /Library/Elastic/Agent | wc -l
    3490
```

#### After this PR
```
$ time sudo ./elastic-agent install -n
Installing in non-interactive mode.
Uninstalling current Elastic Agent... DONE
Copying files................................................. DONE
Installing service... DONE
Starting service... DONE
Elastic Agent has been successfully installed.
sudo ./elastic-agent install -n  0.99s user 6.28s system 52% cpu 13.888 total

$ find . | wc -l
    3464
$ sudo find /Library/Elastic/Agent | wc -l
    3490
```

### MacOS - HDD
TODO

### Linux - SSD
Tested on Google Cloud VM (Ubuntu 22) running SSDs ([balanced persistent disk](https://cloud.google.com/compute/docs/disks#disk-types)).

#### Before this PR
```
$ time sudo ./elastic-agent install -n
Installing in non-interactive mode.Uninstalling current Elastic Agent... DONE
Copying files................................................ DONE
Installing service..... DONE
Starting service... DONE
Elastic Agent has been successfully installed.

real	0m12.150s
user	0m0.003s
sys	0m0.008s

$ find . | wc -l
3448
$ sudo find /opt/Elastic/Agent/ | wc -l
3478
```

#### After this PR

```
$ time sudo ./elastic-agent install -n
Installing in non-interactive mode.
Uninstalling current Elastic Agent... DONE
Copying files...................................... DONE
Installing service.... DONE
Starting service.... DONE
Elastic Agent has been successfully installed.

real	0m9.148s
user	0m0.010s
sys	0m0.000s

$ find . | wc -l
3448
$ sudo find /opt/Elastic/Agent/ | wc -l
3478
```

### Linux - HDD
Tested on Azure Cloud VM (Ubuntu 22) running HDDs ([Standard HDD](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-types#standard-hdds) [LRS](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-redundancy#locally-redundant-storage-for-managed-disks)). 

#### Before this PR
```
$ time sudo ./elastic-agent install -n
Installing in non-interactive mode.Uninstalling current Elastic Agent... DONE
Copying files......................................................................................................................................................................... DONE
Installing service.... DONE
Starting service... DONE
Elastic Agent has been successfully installed.

real	0m45.694s
user	0m0.015s
sys	0m0.005s

$ find . | wc -l
3448
$ sudo find /opt/Elastic/Agent/ | wc -l
3478
```

#### After this PR
```
$ time sudo ./elastic-agent install -n
Installing in non-interactive mode.
Uninstalling current Elastic Agent... DONE
Copying files......................................................................................................................................................................... DONE
Installing service..... DONE
Starting service... DONE
Elastic Agent has been successfully installed.

real	0m45.957s
user	0m0.001s
sys	0m0.019s

$ find . | wc -l
3448
$ sudo find /opt/Elastic/Agent/ | wc -l
3478
```

### Windows - SSD
Tested on 2022 Dell laptop (Windows 11 Pro) running SSDs. 

#### Before this PR
```
PS>  Measure-Command { .\elastic-agent.exe install -n }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 9
Milliseconds      : 616
Ticks             : 96167912
TotalDays         : 0.000111305453703704
TotalHours        : 0.00267133088888889
TotalMinutes      : 0.160279853333333
TotalSeconds      : 9.6167912
TotalMilliseconds : 9616.7912

PS> (Get-ChildItem -Recurse | Measure-Object).Count
3454
PS> cd 'C:\Program Files\Elastic\Agent\'
PS> (Get-ChildItem -Recurse | Measure-Object).Count
3482
```

#### After this PR
```
PS> Measure-Command { .\elastic-agent.exe install -n }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 6
Milliseconds      : 316
Ticks             : 63167305
TotalDays         : 7.3110306712963E-05
TotalHours        : 0.00175464736111111
TotalMinutes      : 0.105278841666667
TotalSeconds      : 6.3167305
TotalMilliseconds : 6316.7305

PS> (Get-ChildItem -Recurse | Measure-Object).Count
3454
PS> cd 'C:\Program Files\Elastic\Agent\'
PS> (Get-ChildItem -Recurse | Measure-Object).Count
3482
```

### Windows - HDD
TODO

## Why is it important?

Better user experience.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] ~I have added an integration test or an E2E test~

### CODE REVIEW

**Review by faec — APPROVED** [review]
Looks good, modulo the question about comments

File: internal/pkg/agent/install/install.go [hunk 1]
```diff
@@ -221,3 +227,35 @@ func verifyDirectory(dir string) error {
 	}
 	return nil
 }
+
+// HasAllSSDs returns true if the host we are on uses SSDs for
+// all its persistent storage; false otherwise or on error
+func HasAllSSDs() bool {
+	block, err := ghw.Block()
+	if err != nil {
+		//fmt.Fprintf(os.Stdout, "HasAllSSDs: ghw.Block() returned error: %s\n", err.Error())
```
[the remark this answers is not in the document]
Comment (ycombinator) [reply internal/pkg/agent/install/install.go:236]: Leftovers, will remove. Thanks!

**Comment by ycombinator** [comment]
Currently waiting on https://github.com/otiai10/copy/pull/123 to be merged, then will update `go.mod` on this PR.

**Comment by ycombinator** [comment]
Hey @michalpristas, I was wondering if you have a Windows machine with an HDD (as opposed to SSD).  @pierrehilbert suggested that you might.  

If you do, would you mind testing the performance of `elastic-agent install` on that machine before and after this PR?  You can follow the same steps that I did for [Windows - SSD](https://github.com/elastic/elastic-agent/pull/3212#issue-1842215062).  

If you don't have such a machine, just let me know.  Thanks either way!

**Comment by ycombinator** [comment]
> Hey @michalpristas, I was wondering if you have a Windows machine with an HDD (as opposed to SSD). @pierrehilbert suggested that you might.
> 
> If you do, would you mind testing the performance of `elastic-agent install` on that machine before and after this PR? You can follow the same steps that I did for [Windows - SSD](https://github.com/elastic/elastic-agent/pull/3212#issue-1842215062) (scroll down a bit).
> 
> If you don't have such a machine, just let me know. Thanks either way!

Chatted with @michalpristas off-PR.  He doesn't have an HDD machine where `elastic-agent` can be installed.

**Comment by ycombinator** [comment]
@cmacknz Worth backporting to `8.11`?

**Comment by cmacknz** [comment]
Missed the original ping here, not a bug fix so we shouldn't have backported.

**Comment by ycombinator** [comment]
> Missed the original ping here, not a bug fix so we shouldn't have backported.

👍 This PR was not backported so we're good.  Thanks!

**Pull request merged**